File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/directory-presence.ts
#!/usr/bin/env bun
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

/**
 * Require specified directories to exist.
 */
export const directoryPresenceRule = (
	ctx: RuleContext,
): HarnessCheckRule => ({
	category: "directoryPresence",
	applies(manifest: HarnessManifest): boolean {
		const section = manifest.directoryPresence;
		if (typeof section !== "object" || section === null) {
			return false;
		}
		const enabled = (section as { enabled?: unknown }).enabled;
		if (enabled === false) {
			return false;
		}
		const entry = ctx.readJsonObject(
			(section as Record<string, unknown>).parameters,
		);
		return ctx.readStringArray(entry.paths).length > 0;
	},

	validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
		const parameters = ctx.readJsonObject(
			ctx.readJsonObject(manifest.directoryPresence).parameters,
		);
		const paths = ctx.readStringArray(parameters.paths);
		return paths.flatMap((path) => {
			if (ctx.isSymlink(path)) {
				return [
					{
						severity: ctx.severityOf(manifest, "directoryPresence"),
						category: "directoryPresence",
						message: `symlink directory is not allowed: ${path}`,
					},
				];
			}
			return ctx.isDirectory(path)
				? []
				: [
						{
							severity: ctx.severityOf(manifest, "directoryPresence"),
							category: "directoryPresence",
							message: `missing directory: ${path}`,
						},
					];
		});
	},
});

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/directory_presence.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require directories exist rule."""

from collections.abc import Iterable

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from .utils import is_safe_directory, severity_for

class DirectoryPresenceRule(HarnessCheckRule):
    """Validate directoryPresence check."""

    category = "directoryPresence"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        return section.get("enabled", True) is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate directoryPresence check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        paths = params.get("paths", [])
        if not isinstance(paths, list):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        template = messages.get("default", "missing directory: {path}")
        return [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                template.format(path=path),
            )
            for path in paths
            if isinstance(path, str) and not is_safe_directory(root / path)
        ]


RULE: HarnessCheckRule = DirectoryPresenceRule()
